cmd/knitgateway/main.go

```diff
@@ -88,19 +90,70 @@ func main() {
 
 	mux := http.NewServeMux()
 	mux.Handle(gateway.AsHandler())
-	corsHandler := cors.AllowAll().Handler(mux)
+	certCapturingHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.TLS != nil && len(r.TLS.PeerCertificates) > 0 {
+			r = r.WithContext(knitgateway.ContextWithClientCert(r.Context(), r.TLS.PeerCertificates[0]))
+		}
+		mux.ServeHTTP(w, r)
+	})
+	// TODO: CORS should not be so lenient unless configured to be so
+	corsHandler := cors.AllowAll().Handler(certCapturingHandler)
 	loggingHandler := knitgateway.NewLoggingHandler(corsHandler, logger)
 	svr := http.Server{
 		Handler:           h2c.NewHandler(loggingHandler, &http2.Server{}),
 		ReadHeaderTimeout: 30 * time.Second,
+		TLSConfig:         config.TLSConfig,
 	}
-	l, err := net.Listen("tcp", config.ListenAddress)
-	if err != nil {
-		fatalf("failed to listen on bind address %s: %v\n", config.ListenAddress, err)
+	group, grpCtx := errgroup.WithContext(ctx)
+	shutdownComplete := make(chan struct{})
+	go func() {
```

Comment:
This can also be part of the group, we can get rid of the extra channel

Reply:
Ah, right, good call. Silly me!

Reply:
Done. While in here and examining the interactions between the goroutines in the group and this shutdown, I decided what this really needed was some signal handling. So I also added a signal handler to initiate proper shutdown.

Reply:
While testing the signal handling, I noticed that it was erroneously creating a gRPC reflection stream per service to download, but the intent was for it to be per backend (so all services for the same backend should use a single stream). So I fixed that, too, while I was in here.